File: 567. Permutation in String/optimised1.py
#!/usr/bin/env python

def checkInclusion(s1: str, s2: str) -> bool:
    n = len(s1)
    m = len(s2)

    if n > m:
        return False

    s1Count = [0] * 26
    s2Count = [0] * 26

    for i in range(n):
        s1Count[ord(s1[i]) - ord('a')] += 1
        s2Count[ord(s2[i]) - ord('a')] += 1

    if s1Count == s2Count:
        return True

    for i in range(n, m):
        s2Count[ord(s2[i]) - ord('a')] += 1
        s2Count[ord(s2[i - n]) - ord('a')] -= 1

        if s2Count == s1Count:
            return True

    return False

print(checkInclusion("ab", "eidbaooo"))
print(checkInclusion("ab", "eidboaoo"))

File: 567. Permutation in String/optimised1.ts
function checkInclusion(s1: string, s2: string): boolean {
	const n = s1.length;
	const m = s2.length;

	if (n > m) return false;

	const s1Count = new Array(26).fill(0);
	const s2Count = new Array(26).fill(0);
	const a = 'a'.charCodeAt(0);

	for (let i = 0; i < n; i++) {
		s1Count[s1.charCodeAt(i) - a]++;
		s2Count[s2.charCodeAt(i) - a]++;
	}

	if (s1Count.join() === s2Count.join()) return true;

	for (let i = n; i < m; i++) {
		s2Count[s2.charCodeAt(i) - a]++;
		s2Count[s2.charCodeAt(i - n) - a]--;

		if (s1Count.join() === s2Count.join()) return true;
	}

	return false;
};

console.log(checkInclusion("ab", "eidbaooo"));
console.log(checkInclusion("ab", "eidboaoo"));
